pages/blog/graphql-authz.mdx

```diff
@@ -0,0 +1,174 @@
+
+Since the executable schema is clean from the authorization logic it simplifies use of the executable schema as a data source. For example, instead of calling a database, the graphql function from graphql.js package could be called, while passing the executable schema as an argument along with graphql query. In such a way the authorization layer is not dependent on the underlying database(s), its architecture, and ORM(s). It is dependent only on the GraphQL schema which is a dependency of GraphQL authorization layer by design.
+
+```typescript
+// using schema as a data source inside pre-execution rule
+
+const CanPublishPost = preExecRule()(async (context, fieldArgs) => {
+  const graphQLResult = await graphql({
+    schema: context.schema,
+    source: `query post { post(id: ${fieldArgs.postId}) { author { id } } }`,
+  });
+
+  const post = graphQLResult.data?.post;
+
+  return post && post.author.id === context.user?.id;
+});
+```
+
+This transforms the technique into an extremely powerful addition, specifically for microservice architectures with centralized gateway-level authorization. The authorization logic could require data points resolved by different microservices. Some data points could be resolved by third-party microservices or APIs. Data can be stored in different databases. Resolving certain data points could be moved to different microservice with different data sources. Using GraphQL schema as a data source, authorization rules simply don’t know all of the details and rely only on GraphQL Schema.
+
+## Reducing database calls with post-execution rules
+
+Querying data sources from authorization rules is powerful but it adds workload to data sources and, in most cases, the data needed to perform authorization is closely related to requested entities. There is a possibility to reduce data sources workload by performing data-dependent authorization logic in the post-execution phase.
+
+With GraphQL AuthZ it’s possible for post-execution rules to require the data, related to the rule target, that is necessary for the rule to work. For example, if the rule is attached to some Object Field it could require any information about sibling fields with their relations. Another example is if the rule is attached to Object Type it could require any fields of the object with their relations.
+
+```typescript
+// creating post-execution rule
+
+const CanReadPost = postExecRule({
+  selectionSet: '{ status author { id } }',
+})(
+  (context, fieldArgs, post, parent) =>
+    post.status === 'public' || post.author.id === context.user?.id
+);
+```
+
+This technique helps to avoid additional calls to data sources by executing authorization logic on top of the entire operation result with the related extra data being populated. Since related data is often stored in the same place it can be fetched from a data source in one call instead of making one call for authorization and one call for actual data populating.
```

Comment:
```suggestion
By using this technique we reduce remote procedure calls to individual data sources by executing authorization logic on top of GraphQL execution result that is enriched with the additional data specified by the authorization rules. Since related data is often stored in the same place it can be fetched from a data source in one roundtrip (via the GraphQL schema) instead of performing one remote procedure call for the authorization and one call for the actual data populating.
```